File: PlaywireObjC/ads/view/banner/AnchoredBannerViewController.m
//
//  AnchoredBannerViewController.m
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 01/06/23.
//  Copyright © 2023 Playwire. All rights reserved.
//

#import "AnchoredBannerViewController.h"
#import <Playwire-Swift.h>

@interface AnchoredBannerViewController () <PWViewAdDelegate>
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (strong, nonatomic) PWBannerViewAnchored *bannerView;
@property (assign, nonatomic) BOOL isBannerAdded;
@end

@implementation AnchoredBannerViewController

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    if (self.isBannerAdded) return;
    [self loadBanner];
}

- (void)loadBanner {
    self.bannerView = [[PWBannerViewAnchored alloc] initWithAdUnitName:self.adUnitName delegate:self];
    
    // Use `[[PWLoadParams new] withTargeting:]` to pass your custom targets to ad request.
    // Use `[[PWLoadParams new] withWidth:]` to pass available width to fill with ad content. Get the width of the device in use, or set your own width if you don’t want to use the full width of the screen.
    // Use `[[PWLoadParams new] withDeviceOrientation:]` to include the orientation of your interface to ad rerquest. Use `.unknown` value to allow SDK to determine the orientation.
//    PWLoadParams *params = [[[[PWLoadParams new]
//                              withTargeting:@{@"age": @"18-32", @"page": @"travel"}]
//                             withWidth: 320]
//                            withDeviceOrientation:UIInterfaceOrientationUnknown];
//     [self.bannerView loadWithParams:params];
    
    // Determine the view width to use for the ad width.
    CGRect frame = UIEdgeInsetsInsetRect(self.view.frame, self.view.safeAreaInsets);
    CGFloat viewWidth = frame.size.width;
    
    PWLoadParams *params = [[PWLoadParams new] withWidth:viewWidth];
    [self.bannerView loadWithParams:params];
    
    self.statusLabel.text = [NSString stringWithFormat: @"⏳ The banner \"%@\" is loading.", self.adUnitName];

}

- (void)addBannerToSuperView {
    if (self.isBannerAdded) return;
    self.isBannerAdded = YES;
    
    // Banner is ready to be added to view hierarchy
    self.bannerView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.bannerView];
    [NSLayoutConstraint activateConstraints:@[
        [self.bannerView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-16],
        [self.bannerView.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.bannerView.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor]
    ]];
}

-(IBAction)refreshAction:(UIButton *)sender {
    // Refresh will start only if the ad unit contains `refresh` object.
    // See logs from `PWNotifier` to track status of refresh.
    
    [self.bannerView refresh];
    
    NSArray *adUnits = [[[PlaywireSDK shared] config] adUnits];
    PWBannerRefresh *refresh;
    for(PWAdUnit *adUnit in adUnits) {
        if([adUnit.name isEqualToString:self.adUnitName]) {
            refresh = adUnit.refresh;
            break;
        }
    }
    
    if (!refresh) {
        self.statusLabel.text = [NSString stringWithFormat:@"⚠️ The banner \"%@\" can't be refreshed manually.\nSee logs to get more details.", self.adUnitName];
        return;
    }
    
    self.statusLabel.text = [NSString stringWithFormat:@"🔄 The banner \"%@\" is refreshing.", self.adUnitName];
}

#pragma mark - PWViewAdDelegate -

- (void)viewAdDidLoad:(PWViewAd * _Nonnull)ad {
    if (!ad.isLoaded) return;
    
    self.statusLabel.text = [NSString stringWithFormat: @"✅ The banner \"%@\" is loaded.", self.adUnitName];
    [self addBannerToSuperView];
}

- (void)viewAdDidFailToLoad:(PWViewAd * _Nonnull)ad {
    self.statusLabel.text = [NSString stringWithFormat: @"❌ Failed to load the banner \"%@\".", self.adUnitName];
}

- (void)viewAdDidRecordClick:(PWViewAd * _Nonnull)ad {}
- (void)viewAdDidRecordImpression:(PWViewAd * _Nonnull)ad {}
- (void)viewAdWillPresentFullScreenContent:(PWViewAd * _Nonnull)ad {}
- (void)viewAdWillDismissFullScreenContent:(PWViewAd * _Nonnull)ad {}
- (void)viewAdDidDismissFullScreenContent:(PWViewAd * _Nonnull)ad {}

@end

File: PlaywireSDKAppsSwift/ads/view/banner/AnchoredBannerViewController.swift
//
//  AnchoredBannerViewController.swift
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 12/20/22.
//  Copyright © 2023 Playwire. All rights reserved.
//

import Playwire
import UIKit

final class AnchoredBannerViewController: UIViewController, AdUnitViewControllerType {
    
    @IBOutlet weak var statusLabel: UILabel!
    
    // The ad unit name, e.g. 'banner-320x50', 'interstitial-home', 'rewarded-coins', etc.
    var adUnitName: String!
    
    private var bannerView: PWBannerViewAnchored!
    private var isBannerAdded: Bool = false
    private var name: String { adUnitName ?? "NULL"}
    
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        guard !isBannerAdded else { return }
        loadBanner()
    }
    
    private func loadBanner() {
        bannerView = PWBannerViewAnchored(adUnitName: adUnitName, delegate: self)
        
        // Use `PWLoadParams().withTargeting()` to pass your custom targets to ad request.
        // Use `PWLoadParams().withWidth()` to pass available width to fill with ad content. Get the width of the device in use, or set your own width if you don’t want to use the full width of the screen.
        // let params = PWLoadParams().withTargeting(
        //  [
        //    "age": "18-32",
        //    "page": "travel"
        //  ]
        // )
        // .withWidth(320)
        // bannerView.load(params: params)
        
        
        // Determine the view width to use for the ad width.
        let frame = { () -> CGRect in
            // Here safe area is taken into account, hence the view frame is used after the view has been laid out.
            if #available(iOS 11.0, *) {
                return view.frame.inset(by: view.safeAreaInsets)
            } else {
                return view.frame
            }
        }()
        let viewWidth = frame.size.width
        
        let params = PWLoadParams()
            .withWidth(viewWidth)
        
        bannerView.load(params: params)
        statusLabel.text = "⏳ The banner \"\(name)\" is loading."
    }
    
    private func addBannerToSuperView() {
        guard !isBannerAdded else { return }
        isBannerAdded = true
        
        // Banner is ready to be added to view hierarchy
        bannerView.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(bannerView)
        NSLayoutConstraint.activate([
            bannerView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -16),
            bannerView.centerXAnchor.constraint(equalTo: view.safeAreaLayoutGuide.centerXAnchor),
            bannerView.leadingAnchor.constraint(greaterThanOrEqualTo: view.safeAreaLayoutGuide.leadingAnchor)
        ])
    }
    
    @IBAction func refreshAction(_ sender: UIButton) {
        // Refresh will start only if the ad unit contains `refresh` object.
        // See logs from `PWNotifier` to track status of refresh.
        
        bannerView.refresh()
        
        let adUnits = PlaywireSDK.shared.config?.adUnits
        let refresh = adUnits?.first { $0.name == adUnitName }?.refresh
        guard refresh != nil else {
            statusLabel.text = "⚠️ The banner \"\(name)\" can't be refreshed manually.\nSee logs to get more details."
            return
        }
        statusLabel.text = "🔄 The banner \"\(name)\" is refreshing."
    }
}

// MARK: - PWViewAdDelegate -
extension AnchoredBannerViewController: PWViewAdDelegate {
    func viewAdDidLoad(_ ad: PWViewAd) {
        guard bannerView.isLoaded else { return }
        
        statusLabel.text = "✅ The banner \"\(name)\" is loaded."
        addBannerToSuperView()
    }
    
    func viewAdDidFailToLoad(_ ad: PWViewAd) {
        statusLabel.text = "❌ Failed to load the banner \"\(name)\"."
    }
    
    func viewAdWillPresentFullScreenContent(_ ad: PWViewAd) {
    }
    
    func viewAdWillDismissFullScreenContent(_ ad: PWViewAd) {
    }
    
    func viewAdDidDismissFullScreenContent(_ ad: PWViewAd) {
    }
    
    func viewAdDidRecordImpression(_ ad: PWViewAd) {
    }
    
    func viewAdDidRecordClick(_ ad: PWViewAd) {
    }
}


